### ISSUE TITLE
[RFC] Modularize DistributedDataParallel

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#37002]
## Summary

This project aims at decomposing existing `DistributedDataParallel` (DDP) implementation into multiple smaller pluggable and customizable building blocks. So that applications can customize DDP to best serve specific applications. More specifically, each type of individual building block exposes a fixed set of APIs but can implement different algorithms. The new DDP constructor takes building blocks to assemble a DDP instance. Existing DDP API can stay intact and internally calls into the new DDP implementation.

This project is motivated by the following requirements:

* Enable new features without polluting existing DDP API
    * Many of the new features would require adding more configurations at initialization. Instead of adding all of them into DDP constructor, these configurations should be self-contained inside the implementation of a building block.
* Make DDP/Reducer code more readable
    * Most of DDP core lives in `reducer.cpp`. It starts from a nicely organized implementation, but as we are gradually adding more features, the growing complexity starts to detrimental to readability.
* Allow applications to customize their DDP by picking and extending sub-components
    * Different applications have vastly different requirements, which we will elaborate in the *Targeted Features* section. Packing everything in one implementation would inevitably leads to unnecessary slowdowns. Instead, applications should be able to just pick the algorithms they need and get rid of those unnecessary complexities. 
* Add C++ DDP API
* Gradually remove legacy features, e.g., `device_ids`, `dim`, `check_reduction`, etc. 
    * `device_ids` is necessary for single-process multi-GPU mode, which is not the recommended mode as GIL contention and the per-iteration scatter and gather could introduce additional overhead. The recommended one process per module replica/device use case does not need `device_ids` or `dim`.
* Gradually move DDP from `torch.nn.parallel` to `torch.distributed` 
    * Currently DDP API lives in `torch.nn.parallel` but the core `Reducer` implementation is in the distributed package, which creates a weird dependency that the single-machine package depends on the distributed package. We can resolve this by moving DDP into the `torch.distributed` package.

## Compositional Building Blocks

We propose to decompose the main logic of `DistributedDataParallel` into the following four building blocks

* **Grad Reader**: responsible for reading grad from the local model, which is done by manipulating DDP hooks, including
    * Where to read the grad from: e.g., `parameter.grad` or from distributed autograd context.
    * When to read grad: e.g., `AccumlateGrad` , forward output tensor, etc.
    * Which grad to expect: e.g., `find_unused_parameters` should skip hooks on unused parameters
* **Grad Bucketer**: responsible for organizing grads into buckets, including
    * How do grads maps to buckets: e.g., static mapping determined at construction time, dynamic mapping based on grad-ready order, etc.
    * Determine when a bucket is ready for communication
    * Grad/bucket compression
* **Comm Scheduler**: responsible for how to communicate buckets, including
    * What comm strategy to use: e.g., `AllReduce` or gossip
    * How to prioritize communication
* **Grad Writer**: responsible for writing the communicated grad back to the local model, including
    * Where to read the grad from: e.g., `parameter.grad` or from distributed autograd context.
    * Trigger post DDP-grad hooks if there are any: e.g., this is useful to overlap optimizer `step` with DDP communication. 

## Targeted New Features

* **Mix DDP with RPC**
    * Use DDP in conjunction with RPC functions.  Currently, DDP hook is only registered to the `AccumulateGrad` function and looks for grad from `param.grad`. To support RPC, DDP need to read grad from and write grad to the distributed autograd context.
    * Compositional building blocks: 
        * Grad Reader
        * Grad Writer
* **Delay Allreduce**
    * There are certain scenarios where we cannot make assumptions on how many times a `AccumulateGrad` hook will be called or we cannot traverse the real autograd graph from the outputs (e.g., with checkpoints + reused layers). For these cases, we can have a slower version of DDP which does not overlap comm with comp, so that DDP should be able to support all use cases that would work with local autograd. (see https://github.com/pytorch/pytorch/pull/35083)
    * Compositional building blocks: 
        * Grad Reader
        * Grad Writer
* **Tune parameter to bucket mapping**
    * This tunes the order of `AllReduce` parameters. Ideally, the `AllReduce` order should be the same as gradient ready order. So that DDP can launch communications as soon as possible. We might have to do some profiling to detect the gradient ready order in the backward pass, and might need to change it over iterations. All processes would need to agree on the mapping, but as the mapping adjustment should occur infrequently, we should be able to afford introducing one more `AllReduce` to achieve consensus. (see https://github.com/pytorch/pytorch/pull/35137)
    * Compositional building blocks: 
        * Grad Bucketer
* **Layer Drop/Skip Comm for Unused Params **
    * Currently, DDP always `AllReduce` all buckets (i.e., all grads) even if some parameters are not used in the forward pass. Some applications intentionally skip layers to speed up training and prevent overfitting. In some other applications, different inputs might reach different part of the model.  Allowing skip unused parameters in the communication as well will help to speed up training. One challenge is that all processes need to have a global consensus on which gradients to skip. (see [paper](https://arxiv.org/pdf/1909.11556.pdf) as one example use case)
    * Compositional building blocks:
        * Grad Reader
        * Grad Bucketer
        * Grad Writer
* **Overlap DDP with optimizer**
    * Existing DDP enforces a hard barrier between backward and optimizer step. However, this is not necessary, as optimizers should be able to start working as soon as some grad is ready. Therefore, we should allow optimizer step to overlap with DDP backward. 
    * Compositional building blocks:
        * Grad Writer
* **Overlap DDP backward with next forward**
    * Description: the reason for computing grad is to use the updated parameter in the next iteration. The next forward does not have to be block until all parameter are updated. Instead, it can start as long as the first layer parameters are ready. So we should allow overlapping DDP backward with next iteration’s forward, and only block forward on a layer if its parameter hasn’t been updated yet.
    * Challenge: API design
    * Compositional building blocks:
        * per sub-module pre-forward hooks (to block forward until parameter is updated, this is already available)
* **Prioritize and preempt allreduce**
    * Grads becomes ready in the backward order, but parameters are consumed by the next forward in the opposite order. If we want to overlap backward with next forward well, we need to assign higher priorities to `AllReduce` operations for first layers. However, as buckets for last layers will be ready first, they will start early. Hence, we need some way to implement *preemption* in allreduce operations.  (see [paper1](https://dl.acm.org/doi/10.1145/3341301.3359642) and [paper2](https://arxiv.org/abs/1905.03960))
    * Compositional building blocks:
        * Comm Scheduler
* **Gossip DDP**
    * Instead of using `AllReduce` to globally sync gradients, it’s proven that gossip and async param sync also works and can potentially lead to considerable training speedups. 
    * Compositional building blocks:
        * Comm Scheduler
* **Gradient Compression**

    * Mixed precision training has been proven to help accelerate training considerably. DDP should support this use case as well by allowing use less bits to store and communicate grads.
    * Compositional building blocks:
        * Grad Bucketer

**Comment by mcarilli** [comment]
Love these!  (Many are already done.)  I would add
- use multi tensor apply to perform bucketing (+ pre-allreduce div) and unbucketing with a single kernel launch, which saves CPU overhead (requires https://github.com/pytorch/pytorch/pull/41554 and a followup PRs with multi tensor mul_ support).  This would also allow dynamically allocating and filling each bucket when that bucket's gradients are ready to kick off.  Creating buckets on-the-fly as needed (instead of pre-allocating all the buckets permanently, as is done today) would reduce the model's memory high water mark.

**Comment by pritamdamania87** [comment]
Regarding `Delay Allreduce`, I'm wondering if we can run in delay allreduce mode only for the first iteration and keep track of how many times each parameter is supposed to be used. Then for subsequent iterations we can probably use this information to run in the normal mode using the information we learnt in the first pass in terms of how many times a parameter can be used. This won't work in the case where there are some conditionals where things might change from iteration to iteration, but I think it would probably serve a lot of use cases well.

**Comment by rohan-varma** [comment]
> Regarding `Delay Allreduce`, I'm wondering if we can run in delay allreduce mode only for the first iteration and keep track of how many times each parameter is supposed to be used. 

Might be my lack of understanding, but was wondering why we have to run in delay allreduce mode to keep track of this? With the current mode where comm/comp are overlapped, won't we be able to get this information by counting how many parameters have their autograd hooks fired?

**Comment by pritamdamania87** [comment]
> Might be my lack of understanding, but was wondering why we have to run in delay allreduce mode to keep track of this? With the current mode where comm/comp are overlapped, won't we be able to get this information by counting how many parameters have their autograd hooks fired?

As per my understanding we mark parameters as "ready" when the autograd hooks fire. If the autograd hooks fire multiple times for the same parameter or not enough times, we probably have asserts in the code that fail. We don't have information upfront about how many times the autograd hook is going to fire before we can mark a parameter as "ready" and I think that's why we need delay allreduce.

**Comment by mrshenli** [comment]
It has been a while, here are some more detailed implementation plans.

## Targeted Features Continued

* **Gossip DDP**
    * Instead of using allreduce to globally sync gradients, it’s proven that gossip and async param sync also works and can potential lead to considerable training speedups.
    * Compositional building blocks:
        * `CommStrategy`
* **Gradient Compression**
    * Compress gradients before communicating them to reduce the communication overhead. This can be done today using the `DDP` communication hook feature. 
    * Compositional building blocks:
        * `CommStrategy`
* **AdaScale**
    * Compute statistics of gradients before and after all_reduce, and use these features to automatically adjust the learning rate as described in [this paper](https://arxiv.org/abs/2007.05105). This can be very useful when scaling to larger numbers of workers/gpus, and could be especially useful when doing elastic training if the number of gpus increases or decreases during training. [Min Xu](https://fb.quip.com/DDSAEAwVAOx)
    * Compositional building blocks:
        * `GradAccessor`



## Composable Engine API

### Requirements:

* The meaning of each configuration can be easily and clearly explained and understood.
* Configurations should be able to mix and match, which allows users to customize the behavior.
* We can carry out existing DDP using the new configurations and APIs. 


The text below calls the new reducer an `Engine`,(Open to better names) as we will see, it behaves like an event-driven engine, regardless of whether the applications use it for gradient reduction or just local logging. The `Engine` only sees the event graph and calls `handle_event` method accordingly. One thing to add is that, this is a synchronized `Engine`, as there won’t be event listening loops or queues and itself does not create additional threads either.  


### Orthogonal Configurations

Based on the use cases we saw so far, users need to configure in 5 orthogonal dimensions. Logically, each of these 5 configurations should be able to mix and match with other configurations. 

* **When** (1) to trigger gradient reduction (`TriggerStrategy`): 
    * DefaultTrigger (trigger DDP logic on every grad ready signal)
    * DelayedTrigger (i.e., delay allreduce)
* **Where** (2) to read and write gradients (`GradAccessor`): (this might not be necessary, as we can detect as of today)
    * DefaultGradAccessor
    * RpcGradAccessor
* **What** (3) to include in the buckets ( `BucketIndexer`)
    * DefaultBucketIndexer
    * IncludeUnused (i.e., `find_unused_parameters`)
    * SkipUnused (layer drop)
* **How** (4) to allocate memory (`BucketAllocator`)
    * DefaultAllocator
    * GradAsBucketView
    * OnDemand (creating buckets on demand)
* **How** (5) to reduce gradient buckets (`CommStrategy`):
    * AllReduceComm
    * CompressionComm
    * GossipComm

### Composition API Examples

#### Representing Legacy `torch.nn.parallel.DistributedDataParallel`


```python
class DistributedDataParallel:

    def __init__(
        self, 
        module,
        device_ids,
        ...
        bucket_cap_mb, 
        find_unused_parameters, 
        gradient_as_bucket_view
    ):
        ...
        # device_ids for SPMD use cases will be retired soon
        if len(device_ids) == 1:
            bucket_indexer = IncludeUnused if \
                find_unused_parameters else DefaultBucketIndexer()
            bucket_allocator = GradAsBucketView(bucket_cap_mb) if \
                gradient_as_bucket_view else DefaultAllocator(bucket_cap_mb)
        

            self.engine = Engine(
                gradient_accessor=DefaultGradAccessor(),  # optional
                trigger_strategy=DefaultTrigger(),  # optional
                comm_strategy=AllReduceComm(),
                bucket_indexer=bucket_indexer,
                bucket_allocator=bucket_allocator
            )
        
            self.engine.prepare_for_module(module)
            
        else:
            self.reducer = Reducer()
        
        
    def forward(self, *inputs, **kwargs):
        self.engine.pre_forward(*inputs, **kwargs)
        
        ...
        
        self.engine.post_forward(outputs)
        
        
    def _register_comm_hook(self, state, hook):
        self.engine._set_comm_strategy(PythonCommHookStrategy(state, hook))
        
        
    def join(self, divide_by_initial_world_size, enable):
        ...
        while not all_procs_joined:
            ...
            self.engine.dummy_backward()
            ...


```

#### Creating new DistributedDataParallel

Existing DDP lives in `torch.nn.parallel`, which is a little weird as `torch.distributed` is clearly a better package path for it. We can take this chance to create a new `DistributedDataParallel` and do the package migration. The old one will still be there, but will be in a maintenance mode. For the new one, we probably need a new package, as there will be other types of data parallel or utils introduced in the future. The code below uses `spmd` (single-program, multi-data) as the package name. Please do shout out if there are better names.

```python
# new torch.distributed.spmd.DistributedDataParallel
class DistributedDataParallel:
    def __init__(self, module, engine=None):
        self.reducer = reducer if reducer else DefaultReducer()
        
# user code
engine= Engine(
    gradient_accessor=RpcGradAccessor(),
    trigger_strategy=DelayedTrigger(), 
    bucket_indexer=SkipUnused(),
    bucket_allocator=GradAsView()
)
ddp = DistributedDataParallel(module, engine)
```


## Extension API

Besides allowing users to compose DDP in arbitrary way they want, it would also be great to support extending it with Python user code, so that power users can conveniently try out new algorithms. As we will see later in the implementation section, from the engine's perspective, there is no differences from the 4 components (except `GradAccessor`), as they are all just event handlers. `GradAccessor` is more like a utility. So we can give greater flexibility to the extension API. Users can build the engine from whatever event handlers they provide. E.g.:


```python
engine = build_engine_from(
    DefaultGradAccessor(),
    [
        MyTrigger(), 
        MyBucketIndexer(200), 
        OnDemandBucketAllocator(),
        AllReduceComm(),
        MyLogger(),
        MyWhateverEventHandler()
    ]
)

ddp = torch.distributed.simd.DistributedDataParallel(module, engine)
```

In the above example, users can create a `MyLogger` event handler to access all DDP internal events and signals, which is not visible to users in today’s DDP.  

Matured, well-tested, and widely-used components should be added to our C++ library and pybind to Python API, if adding them in C++ leads to considerably higher efficiency. But users should be able to experiment with new ideas by writing pure Python code.


## Implementation

Let’s first take a look at what are the use cases we support and plan to support, and what they do when different event occurs. First 8 rows are existing features.


|scenarios vs events	|PREPARE MODULE	|PRE FORWARD	|POST FORWARD	|PRE BACKWARD	|GRAD READY	|BUCKET READY	|COMM DONE	|POST BACKWARD	|
|---	|---	|---	|---	|---	|---	|---	|---	|---	|
|default	|buckets, hooks	|	|	|	|copy to bucket	|trigger comm	|copy to grad	|wait for all comm	|
|find unused	|buckets, hooks	|	|detect unused params	|	|copy to bucket, mark unused as ready	|trigger comm	|copy to grad	|wait for all comm	|
|delay allreduce	|buckets	|	|	|	|	|	|	|reduce all grad	|
|rebuilt buckets	|buckets, hooks	|rebuilt buckets	|	|	|copy to bucket	|trigger comm	|copy to grad	|wait for all comm	|
|grad as view	|buckets, hooks	|	|	|	|copy to bucket, set view	|trigger comm	|	|wait for all comm	|
|RPC + DDP	|buckets, hooks	|	|	|	|copy to bucket	|trigger comm	|copy to RPC ctx	|wait for all comm	|
|no sync	|buckets, hooks	|skip comm	|	|	|copy to bucket	|trigger comm	|copy to grad	|wait for all comm	|
|join	|buckets, hooks	|	|	|similate comm	|	|	|	|	|
|deepspeed	|buckets	|load parameters	|drop params and checkpoint activations	|load param, restore activations	|copy to bucket	|trigger comm	|copy to grad, or discard	|drop non-owner params and grads	|
|gossip	|buckets, hooks	|	|	|	|copy to bucket	|trigger comm	|unclear	|unclear	|
|Layer Drop	|hooks	|maybe sync schedule	|	|ignore dropped ones	|copy to bucket	|trigger comm	|copy to grad	|wait for all comm	|
|On demand buckets	|hooks	|	|	|	|record index	|build bucket, trigger comm	|copy to grad, delete bucket	|wait for all comm	|


If we look at each scenario as one entity, it can be quite complex. But if we look at what they do to respond to each specific event, the logic become a lot more simpler. The events can be categorized into three types


* Type 1 Events: exist in user code 
    * PREPARE_MODULE: produced by reducer
    * PRE_FORWARD: produced by reducer
    * POST_FORWARD: produced by reducer
* Type 2 Events: exist without DDP
    * GRAD_READY (by inserting autograd hook): usually produced by trigger
    * PRE_BACKWARD (by inserting a phony layer): usually produced by trigger
    * POST_BACKWARD (by inserting a phony layer): usually produced by trigger
        * `autograd::Engine::queue_callback()` does not work for reentrant backward, hence need a phony layer here as well.
* Type 3 Events: DDP-specific
    * BUCKET_CONTENT_READY: usually produced by bucket indexer
    * BUCKET_TENSOR_READY: usually produced by bucket allocator
    * COMM_DONE: usually produced by comm


There can be new events and new scenarios in the future, but the `Engine` itself can stay stable and light-weight, where new logics goes into new event handling components. The `Engine` will take a list of `EventHandler` objects, check whether the events they produce and consume are compatible, and trigger Type 1 events. That’s all an engine will do, and everything else is composable. 


```python
class EventType(Enum):
    PREPARE_MODULE = 1
    PREPARE_FORWARD = 2
    POST_FORWARD = 3
    PRE_BACKWARD = 4
    GRAD_READY = 5
    BUCKET_CONTENT_READY = 6
    BUCKET_TENSOR_READY = 7
    COMM_DONE = 8
    POST_BACKWARD = 9

```

```python
class Frequency(Enum):
    EXACTLY_ONCE = 1 # this event occurs exactly once per iteration
    AT_MOST_ONCE = 2 # this event occurs at most once per iteration
```

```python
class EventSchema:
    type: EventType
    freq: Frequency
    
```

```python
class EventHandler:
    @staticmethod
    def ingress_events() -> List[EventSchema]:
        # returns a list of event types it will consume
        pass
        
    @staticmethod
    def egress_events() -> List[EventSchema]:
        # returns a list of event types it will produce
        pass
        
    @staticmethod
    def handle_event(event, ctx) -> List[Future[Event]]:
        # handles events
        """
        Arguments:
            event (Event): it carries type info and event-specific
                data objects.
            ctx (EngineContext): global context shared across all 
                event handlers. It also exposes access to 
                GradAccessor.
        """
        pass
```


```python
class Engine:
    # using Python pseudo code as an example. Both Engine
    # and Event-releated types will be implemented in C++
    def __init__(self, event_handlers: List[EventHandler]):
        # use ingress_events and egress_events from the 
        # EventHandlers to build a graph, and verify that:
        # 1. each type of event is only produced by exactly one 
        #    EventHandler instance
        # 2. Event frequency of producer and consumer must match
        #    (We can relax this in the future if necessary).
        # 3. The graph must be a connected graph, with only
        #    type I events as ingress edges.
        
    def prepare_for_module(self):
        process_event(torch.futures.Future(PrepareForModuleEvent()))
        
    def pre_forward(self):
        process_event(torch.futures.Future(PreForwardEvent())) 
        
    def post_forward(self):
        process_event(torch.futures.Future(PostForwardEvent())) 
        
    def process_event(fut_event: Future[Event]):
        def cb(fut):
            event = fut.result()
            for consumer in self.event_consumers[event.type]:
                futs = consumer.handle_event(event, self.get_ctx(consumer))
                for fut in futs:
                    self.process_event(fut)
        fut_event.then(cb)
```




When users would like to experiment new ideas, they can implement new Python `EventHandler` as they do today for autograd functions. But all native `EventHandler` we provide should be implemented in C++ for performance reasons.


### Example Event Handlers


**Trigger**


```python
class DefaultTrigger(EventHandler):
    @staticmethod
    def schema():
        return ...
        
    @staticmethod
    def ingress_events():
        return [PREPARE_MODULE]
      
    @staticmethod
    def egress_events():
        return [GRAD_READY]
      
    @staticmethod
    def handle_event(event, ctx) -> List[Future[Event]]:
        if event == PREPARE_MODULE:
            # register hooks on event.module
```



```python
class DelayedTrigger(EventHandler):
    @staticmethod
    def ingress_events():
        return [PREPARE_MODULE, PRE_FORWARD]
        
    @staticmethod
    def egress_events():
        return [GRAD_READY, POST_BACKWARD]
        
    @staticmethod
    def handle_event(event, ctx):
        if event == PREPARE_MODULE:
            # keep of reference of all params
        elif event == PRE_FORWARD
            # insert a phony layer, this phony layer will
            # generate one POST_BACKWARD event and all
            # GRAD_READY events
            
```


**BucketIndexer**


```python
class DefaultBucketIndexer(EventHandler):
    @staticmethod
    def ingress_events():
        return [GRAD_READY, COMM_DONE]
        
    @staticmethod
    def egress_events():
        return [BUCKET_CONTENT_READY]
        
    @staticmethod
    def handle_event(event, ctx):
        if event == GRAD_READY:
            # generate BUCKET_CONTENT_READY accordingly
        elif event == COMM_DONE:
            # copy bucket content back to grad
                                           
```




```python
class IncludeUnsedBucketIndexer(EventHandler):
    @staticmethod
    def ingress_events():
        return [POST_FORWARD, GRAD_READY, COMM_DONE, BACKWARD_DONE]
        
    @staticmethod
    def egress_events():
        return [BUCKET_CONTENT_READY]
        
    @staticmethod
    def handle_event(event, ctx):
        if event == POST_FORWARD:
            # find unused parameters
        elif event == GRAD_READY:
            # generate BUCKET_CONTENT_READY accordingly
        elif event == COMM_DONE:
            # copy bucket content back to grad
        elif event == BACKWARD_DONE:
            # block until all ready buckets are communicated
```



**BucketAllocator**



```python
class DefaultBucketAllocator(EventHandler):
    @staticmethod
    def ingress_events():
        return [PREPARE_MODULE, BUCKET_CONTENT_READY]
        
    @staticmethod
    def egress_events():
        return [BUCKET_TENSOR_READY]
        
    @staticmethod
    def handle_event(event, ctx):
        if event == PREPARE_MODULE:
            # create buckets
        elif event == BUCKET_CONTENT_READY:
            # copy grad to buckets
            # write BUCKET_TENSOR_READY event to ctx.event_hub
```



```python
class GradAsViewBucketAllocator(EventHandler):
    @staticmethod
    def ingress_events():
        return [PREPARE_MODULE, BUCKET_CONTENT_READY]
        
    @staticmethod
    def egress_events():
        return [BUCKET_TENSOR_READY]
        
    @staticmethod
    def handle_event(event, ctx):
        if event == BUCKET_CONTENT_READY:
            # create a bucket and let grad points to views of the bucket offset
            # write BUCKET_TENSOR_READY event to ctx.event_hub
```


**Comm**



```python
class AllReduceComm(EventHandler):
    @staticmethod
    def ingress_events():
        return [BUCKET_TENSOR_READY, BACKWARD_DONE]
        
    @staticmethod
    def egress_events():
        return [COMM_DONE]
            
    @staticmethod
    def handle_event(event, ctx):
        if event == BUCKET_TENSOR_READY:
            # send tensor and generate 
        elif event == BACKWARD_DONE:
            # wait for all pending works and generate COMM_DONE accordingly
```

**Comment by sharvil** [comment]
Following up from https://discuss.pytorch.org/t/custom-gradient-averaging-with-ddp/110926/.

It would be nice to have a client choose which reduction op to perform on gradients. In my case, I have variable and uneven batch sizes on each replica and need to compute the average by *global* batch size.

If I could tell the reducer I wanted a SUM reduction, I can divide by the global batch size myself and get the desired result. It would be even better if I could specify a callback so I can do the allreduce and divide myself in a single step (which, in my case, would also allow for the optimizer to run concurrently with the backward step as buckets become ready).

**Comment by SciPioneer** [comment]
> Following up from https://discuss.pytorch.org/t/custom-gradient-averaging-with-ddp/110926/.
> 
> It would be nice to have a client choose which reduction op to perform on gradients. In my case, I have variable and uneven batch sizes on each replica and need to compute the average by _global_ batch size.
> 
> If I could tell the reducer I wanted a SUM reduction, I can divide by the global batch size myself and get the desired result. It would be even better if I could specify a callback so I can do the allreduce and divide myself in a single step (which, in my case, would also allow for the optimizer to run concurrently with the backward step as buckets become ready).

Just saw this comment. it seems that you can develop a customized DDP communication hook, which should be very similar to [allreduce_hook](https://github.com/pytorch/pytorch/blob/master/torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py#L21), by using a different denominator. DDP communication hook is already supported in [stable release](https://pytorch.org/docs/stable/ddp_comm_hooks.html).
